src/main/scala/za/co/absa/commons/s3/S3Location.scala

```diff
@@ -49,10 +53,28 @@ object SimpleS3Location {
     }
 
     def isValidS3Path: Boolean = S3LocationRx.pattern.matcher(path).matches
-  }
 
-  def apply(path: String): SimpleS3Location = {
-    path.toSimpleS3Location.getOrElse(throw new IllegalArgumentException(s"Could not parse S3 location from $path!"))
+    def withoutTrailSlash: SimpleS3Location = {
+      val parsedS3Location = SimpleS3Location(path)
+
+      if (parsedS3Location.path.endsWith("/")) parsedS3Location.copy(path=parsedS3Location.path.dropRight(1))
+      else parsedS3Location
+    }
+
+    def withTrailSlash: SimpleS3Location = {
+      val parsedS3Location = SimpleS3Location(path)
+
+      if (parsedS3Location.path.endsWith("/")) parsedS3Location
+      else {
+        val lastPartOfS3Path = parsedS3Location.path.split("/").last
+        if (lastPartOfS3Path.contains("."))
```

Comment:
This is not foolproof - files do not need to contain `.`s.
I know that there is no other information available in the instance, though, so perhaps there is no better way other than performing this check at all (probably not desired by the requested use case)

Reply:
Yeah I know, it's more like a 'best effort' kind of a thing. If there is no dot in the path, then it still can be a file, but hopefully not vice versa, i.e. if there is a dot then it's a path.

But when I think about it, we can push this logic outside of this function, so that it's a responsibility of a caller to handle this. And thus this function would just always add a trailing slash if there is not one already. What do you think?

Reply:
I figured it was a best-effort basis.

Sure, pushing to another function or just accompanying with an explanatory comment could be fine.

>  If there is no dot in the path, then it still can be a file, but hopefully not vice versa, i.e. if there is a dot then it's a path.
Strictly speaking, I would not say either ((no) presence of `.` in the path does not prove being a file or not.